### PR TITLE
Fix inspector scrollbars.

### PR DESCRIPTION
**Pull request by jacob314**
Inspector rows are now aware of how much space they have in the viewable window so that we can do a better job highlighting rows like this:
Screen Shot 2021-02-17 at 12 33 26 PM
fixing 
https://github.com/flutter/devtools/issues/2684

and text truncates when it reaches the end of the visible area like this:
Screen Shot 2021-02-17 at 1 06 03 PM

As a followup I'd like to polish how the erros are shown in the tree as.

Mostly fixes https://github.com/flutter/devtools/issues/2626
as we now truncate long lines with ... instead of broken wrapping.

Fixes 
https://github.com/flutter/devtools/issues/2674
but adding keys so inspector state objects are reused better.

### CODE REVIEW

File: packages/devtools_app/lib/src/inspector/inspector_screen.dart [hunk 1]
```diff
@@ -69,6 +69,9 @@ class InspectorScreenBodyState extends State<InspectorScreenBody>
   bool get enableButtons =>
       actionInProgress == false && connectionInProgress == false;
 
+  static const summaryTreeKey = Key('Summary Tree');
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/inspector_screen.dart:72]: This helps ensure inspector tree state widgets are reused when they should which avoids a known bug in the flutter scrollbars as well as helping performance a little.

File: packages/devtools_app/lib/src/inspector/inspector_tree_flutter.dart [hunk 2]
```diff
@@ -406,54 +426,71 @@ class _InspectorTreeState extends State<InspectorTree>
...
-                      final InspectorTreeRow row =
-                          controller.root?.getRow(index);
-                      final inspectorRef = row.node.diagnostic?.valueRef?.id;
-                      return _InspectorTreeRowWidget(
-                        key: PageStorageKey(row?.node),
-                        inspectorTreeState: this,
-                        row: row,
-                        error:
-                            widget.widgetErrors != null && inspectorRef != null
+    return LayoutBuilder(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/inspector/inspector_tree_flutter.dart:435]: make sure to view this change with hide whitespace diffs.

File: packages/devtools_app/lib/src/inspector/inspector_tree_flutter.dart [hunk 3]
```diff
@@ -340,22 +354,28 @@ class _InspectorTreeState extends State<InspectorTree>
     currentAnimateTarget = null;
   }
 
+  double get safeViewportHeight {
+    return _scrollControllerY.hasClients
+        ? _scrollControllerY.position.viewportDimension
+        : 1000.0;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/inspector/inspector_tree_flutter.dart:360]: put this in a const with a name that describes what it is

File: packages/devtools_app/lib/src/ui/utils.dart [hunk 4]
```diff
@@ -109,3 +110,96 @@ TextSpan truncateTextSpan(TextSpan span, int length) {
...
+          // have to implement our own ScrollbarPainter instead of being able
+          // to use the existing Scrollbar widget.
+          setState(() {});
+        }
+      });
+    }
+    return AnimatedBuilder(
+      animation: widget.offsetController,
+      builder: (context, child) {
+        // Compute a delta so that
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/ui/utils.dart:173]: nit: incomplete comment

File: packages/devtools_app/lib/src/ui/utils.dart [hunk 5]
```diff
@@ -109,3 +110,96 @@ TextSpan truncateTextSpan(TextSpan span, int length) {
...
+            : Offset(0.0, delta);
+        return Transform.translate(
+          offset: offset,
+          child: Scrollbar(
+              isAlwaysShown: widget.isAlwaysShown,
+              controller: widget.controller,
+              child: Transform.translate(
+                offset: -offset,
+                child: child,
+              )),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/ui/utils.dart:199]: nit: trailing comma

**Review by kenzieschmoll — APPROVED** [review]
LGTM with a few nits

File: packages/devtools_app/lib/src/inspector/inspector_tree_flutter.dart (hunk 3 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/inspector/inspector_tree_flutter.dart:360]: done

File: packages/devtools_app/lib/src/ui/utils.dart (hunk 4 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/ui/utils.dart:173]: done

**Pull request merged**